### PR TITLE
Update Go SDK generator to support k8s

### PR DESCRIPTION
**Pull request by lblackstone**
- Add GoInfo struct to improve control of package imports and names
- Fix import recursion bug
- Add support for aliased imports

### CODE REVIEW

File: pkg/codegen/go/gen.go [hunk 1]
```diff
@@ -739,51 +749,72 @@ func (pkg *pkgContext) genInitFn(w io.Writer, types []*schema.ObjectType) {
 	fmt.Fprintf(w, "}\n")
 }
 
-func (pkg *pkgContext) getTypeImports(t schema.Type, imports stringSet) {
+// importBasePath converts from a repository URL to an import path.
```
Comment (pgavlin) [inline pkg/codegen/go/gen.go]: In addition to (or instead of) this function, we should add the import base path to the package-level Go info.

File: pkg/codegen/go/gen.go [hunk 2]
```diff
@@ -809,9 +840,19 @@ func (pkg *pkgContext) genHeader(w io.Writer, goImports []string, importedPackag
...
 	if len(importedPackages) > 0 {
 		for k := range importedPackages {
-			imports = append(imports, k)
+			imports = append(imports, fmt.Sprintf("%s", k))
 		}
 		sort.Strings(imports)
+
+		// Match the end of an import path that can be mapped to an alias.
+		//   "github.com/pulumi/pulumi-kubernetes/sdk/go/kubernetes/meta/v1" -> meta/v1
+		re := regexp.MustCompile(`\w+/\w+$`)
```
Comment (pgavlin) [inline pkg/codegen/go/gen.go]: Instead of this, we should normalize the module -> alias map to a package import -> alias map. That will avoid the need for this kind of matching, which won't work for packages that don't match this pattern.

File: pkg/codegen/go/gen.go [hunk 3]
```diff
@@ -826,7 +867,11 @@ func (pkg *pkgContext) genHeader(w io.Writer, goImports []string, importedPackag
 			if i == "" {
 				fmt.Fprintf(w, "\n")
 			} else {
-				fmt.Fprintf(w, "\t\"%s\"\n", i)
+				if strings.Contains(i, `"`) { // Imports with aliases already include quotes.
```
Comment (pgavlin) [inline pkg/codegen/go/gen.go]: I think you're normalizing everything to include quotes earlier; this shouldn't be necessary.

Comment (lblackstone) [reply pkg/codegen/go/gen.go]: Before this change, aliased imports would show up like this:
```
"metav1 "github.com/pulumi/pulumi-kubernetes/sdk/go/kubernetes/meta/v1""
```

Comment (pgavlin) [reply pkg/codegen/go/gen.go]: Yeah, I gotcha--my point was that given the code [here](https://github.com/pulumi/pulumi/pull/4096/files#diff-b02dc03ec48c179b1b6609cd77eb12a2R835), I think this call to `strings.Contains` will always return `true`, so we can simplify this to `fmt.Fprintf(w, "\t%s\n", i)`

File: pkg/codegen/go/gen.go [hunk 4]
```diff
@@ -879,20 +912,56 @@ func (pkg *pkgContext) genConfig(w io.Writer, variables []*schema.Property) erro
...
 func GeneratePackage(tool string, pkg *schema.Package) (map[string][]byte, error) {
+	var goInfo GoInfo
+	err := json.Unmarshal(pkg.Language["go"], &goInfo)
+	if err != nil {
+		return nil, err
+	}
+
+	// importBasePath converts from a repository URL to an import path.
+	//   https://github.com/pulumi/pulumi-kubernetes -> github.com/pulumi/pulumi-kubernetes/sdk/go/kubernetes
+	importBasePath := func() string {
```
Comment (lblackstone) [inline pkg/codegen/go/gen.go]: It might make more sense to put this in the GoInfo rather than computing it.

Comment (pgavlin) [reply pkg/codegen/go/gen.go]: Yes--that was what I was suggesting :)

File: pkg/codegen/go/gen.go (hunk 3 above)
Comment (lblackstone) [reply pkg/codegen/go/gen.go]: Double checked, and think it's correct as is. Aside from the aliased imports, which have quotes in the string already, all of the other imports (goImports + unaliased imports) don't already include quotes.

File: pkg/codegen/go/gen.go [hunk 5]
```diff
@@ -809,9 +832,15 @@ func (pkg *pkgContext) genHeader(w io.Writer, goImports []string, importedPackag
 	var imports []string
 	if len(importedPackages) > 0 {
 		for k := range importedPackages {
-			imports = append(imports, k)
+			imports = append(imports, fmt.Sprintf("%s", k))
```
Comment (pgavlin) [inline pkg/codegen/go/gen.go]: Why is this change necessary?

File: pkg/codegen/go/gen.go (hunk 3 above)
Comment (pgavlin) [reply pkg/codegen/go/gen.go]: Yeah, I see that I misread the change I pointed at earlier :)

File: pkg/codegen/go/gen.go [hunk 6]
```diff
@@ -879,20 +912,53 @@ func (pkg *pkgContext) genConfig(w io.Writer, variables []*schema.Property) erro
 	return nil
 }
 
+// GoInfo holds information required to generate the Go SDK from a schema.
+type GoInfo struct {
+	// Base path for package imports
+	//
+	//    github.com/pulumi/pulumi-kubernetes/sdk/go/kubernetes
+	ImportBasePath string `json:"importBasePath"`
```
Comment (pgavlin) [inline pkg/codegen/go/gen.go]:
```suggestion
	ImportBasePath string `json:"importBasePath,omitempty"`
```

File: pkg/codegen/go/gen.go (hunk 5 above)
Comment (lblackstone) [reply pkg/codegen/go/gen.go]: It's apparently not. Must have been leftover from a previous change.
